gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireContentsProcedure.java

```diff
@@ -0,0 +1,292 @@
+                DataTypes.createArrayType(DataTypes.StringType),
+                true,
+                Metadata.empty())
+          });
+
+  public enum FileType {
+    ICEBERG_MANIFEST,
+    ICEBERG_MANIFESTLIST,
+    DATA_FILE
+  }
+
+  // columns from IdentifiedResultsRepo
+  private static final String COL_CONTENT_ID = "contentId";
+  private static final String COL_METADATA_LOCATION = "metadataLocation";
+  private static final String COL_SNAPSHOT_ID = "snapshotId";
+
+  // columns used for intermediate transforms
+  private static final String COL_EXPIRED_FILES_ARRAY = "expiredFilesArray";
+  private static final String COL_EXPIRED_FILES_WITH_TYPE = "expiredFilesWithType";
+  private static final String COL_EXPIRED_FILES_AND_TYPE = "expiredFilesAndType";
+  private static final String COL_EXPIRED_FILES = "expiredFiles";
+  private static final String COL_EXPIRED_FILES_TYPE = "expiredFilesType";
+  private static final String COL_EXPIRED_FILES_COUNT = "expiredFilesCount";
+  private static final String COL_EXPIRED_FILES_LIST = "expiredFilesList";
+
+  public ExpireContentsProcedure(TableCatalog currentCatalog) {
+    super(currentCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public String description() {
+    return String.format(
+        "Expires the Iceberg snapshots that are collected by '%s' procedure.",
+        IdentifyExpiredContentsProcedure.PROCEDURE_NAME);
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow internalRow) {
+    String gcCatalogName = internalRow.getString(0);
+    String gcOutputBranchName = internalRow.getString(1);
+    String gcOutputTableIdentifier = internalRow.getString(2);
+    Map<String, String> nessieClientConfig = new HashMap<>();
+    MapData map = internalRow.getMap(3);
+    for (int i = 0; i < map.numElements(); i++) {
+      nessieClientConfig.put(
+          map.keyArray().getUTF8String(i).toString(), map.valueArray().getUTF8String(i).toString());
+    }
+    String runId = !internalRow.isNullAt(4) ? internalRow.getString(4) : null;
+    boolean dryRun = !internalRow.isNullAt(5) && internalRow.getBoolean(5);
+
+    IdentifiedResultsRepo identifiedResultsRepo =
+        new IdentifiedResultsRepo(
+            spark(), gcCatalogName, gcOutputBranchName, gcOutputTableIdentifier);
+
+    if (runId == null) {
+      runId = getLatestCompletedRunID(gcOutputTableIdentifier, identifiedResultsRepo);
+    }
+
+    try (FileIO fileIO = getFileIO(nessieClientConfig)) {
+      Dataset<Row> expiredContentsDF = getExpiredContents(runId, identifiedResultsRepo, fileIO);
+      if (!dryRun) {
+        expiredContentsDF =
+            expiredContentsDF.map(
+                new DeleteFunction(fileIO), RowEncoder.apply(expiredContentsDF.schema()));
+      }
+      List<Row> rows = expiredContentsDF.collectAsList();
```

Comment:
This still looks unnecessary. Why not `return expiredContentsDF.map().collect()` with a `MapFunction` ??